server/src/main/scala/actors/ShiftsActorBase.scala

```diff
@@ -59,52 +65,66 @@ class ShiftsActor(now: () => SDateLike) extends ShiftsActorBase(now) {
 
-  def initialState = ShiftsState(ShiftAssignments.empty)
+  def initialState: ShiftAssignments = ShiftAssignments.empty
 
-  val snapshotInterval = 1
+  val snapshotInterval = 250
   override val snapshotBytesThreshold: Int = oneMegaByte
 
   import ShiftsMessageParser._
 
-  override def stateToMessage: GeneratedMessage = ShiftStateSnapshotMessage(staffAssignmentsToShiftsMessages(state.shifts, now()))
+  override def stateToMessage: GeneratedMessage = ShiftStateSnapshotMessage(staffAssignmentsToShiftsMessages(state, now()))
 
-  def updateState(shifts: ShiftAssignments): Unit = state = state.updated(data = shifts)
+  def updateState(shifts: ShiftAssignments): Unit = state = shifts
 
   def onUpdateState(data: ShiftAssignments): Unit = {}
 
   def processSnapshotMessage: PartialFunction[Any, Unit] = {
-    case snapshot: ShiftStateSnapshotMessage => state = ShiftsState(shiftMessagesToStaffAssignments(snapshot.shifts))
+    case snapshot: ShiftStateSnapshotMessage => state = shiftMessagesToStaffAssignments(snapshot.shifts)
   }
 
   def processRecoveryMessage: PartialFunction[Any, Unit] = {
-    case shiftsMessage: ShiftsMessage =>
-      log.info(s"Recovery: ShiftsMessage received with ${shiftsMessage.shifts.length} shifts")
-      val shifts = shiftMessagesToStaffAssignments(shiftsMessage.shifts)
-      updateState(shifts)
+    case ShiftsMessage(shiftMessages, _) =>
+      log.info(s"Recovery: ShiftsMessage received with ${shiftMessages.length} shifts")
+      val shiftsToRecover = shiftMessagesToStaffAssignments(shiftMessages)
+      val updatedShifts = applyUpdatedShifts(state.assignments, shiftsToRecover.assignments)
+      purgeExpiredAndUpdateState(ShiftAssignments(updatedShifts))
   }
 
   def receiveCommand: Receive = {
     case GetState =>
       log.info(s"GetState received")
-      sender() ! state.shifts
+      sender() ! state.purgeExpired(expireBefore)
+
+    case UpdateShifts(shiftsToUpdate) =>
+      val updatedShifts = applyUpdatedShifts(state.assignments, shiftsToUpdate)
+      purgeExpiredAndUpdateState(ShiftAssignments(updatedShifts))
+
+      val createdAt = now()
+      val shiftsMessage = ShiftsMessage(staffAssignmentsToShiftsMessages(ShiftAssignments(shiftsToUpdate), createdAt), Option(createdAt.millisSinceEpoch))
+      persistAndMaybeSnapshot(shiftsMessage)
 
-    case shiftStaffAssignments: ShiftAssignments if shiftStaffAssignments != state.shifts =>
-      updateState(shiftStaffAssignments)
-      onUpdateState(shiftStaffAssignments)
+      sender() ! UpdateShiftsAck(shiftsToUpdate)
 
-      log.info(s"Fixed points updated. Saving snapshot")
-      val snapshotMessage = ShiftStateSnapshotMessage(staffAssignmentsToShiftsMessages(state.shifts, now()))
-      saveSnapshot(snapshotMessage)
+    case SetShifts(newShiftAssignments) =>
+      if (newShiftAssignments != state) {
+        log.info(s"Replacing shifts state with $newShiftAssignments")
+        updateState(ShiftAssignments(newShiftAssignments))
```

Comment:
Isn't updateState called in the line below?

Reply:
good catch!